packages/frontend/app/services/channels/channel-verifier.ts

```diff
@@ -0,0 +1,128 @@
+import Service, { inject as service } from '@ember/service';
+import ChannelContextChain from 'emberclear/models/channel-context-chain';
+import VoteChain, { VOTE_ACTION } from 'emberclear/models/vote-chain';
+import VoteVerifier from 'emberclear/services/channels/vote-verifier';
+import Identity from 'emberclear/models/identity';
+import { identityEquals, identityArrayContains } from 'emberclear/utils/identity-comparison';
+
+export default class ChannelVerifier extends Service {
+  @service('channels/vote-verifier') voteVerifier!: VoteVerifier;
+
+  async isValidChain(channel: ChannelContextChain): Promise<boolean> {
+    if (channel.previousChain === undefined) {
+      return this.isValidSingleChain(channel);
+    }
+    if (
+      !(await this.isValidChain(channel.previousChain)) ||
+      !(await this.voteVerifier.verify(channel.supportingVote)) ||
+      !this.checkVoteCompletedPositive(channel.supportingVote, channel.previousChain.admin)
+    ) {
+      return false;
+    }
+
+    switch (channel.supportingVote.action) {
+      case VOTE_ACTION.ADD:
+        return this.checkAdd(channel);
+      case VOTE_ACTION.PROMOTE:
+        return this.checkPromote(channel);
+      case VOTE_ACTION.REMOVE:
+        return this.checkRemove(channel);
+      default:
+        return false;
+    }
+  }
+
+  private isValidSingleChain(channel: ChannelContextChain): boolean {
+    let isInvalidChain =
+      channel.admin === undefined ||
+      channel.members.length !== 1 ||
+      !identityEquals(channel.members.objectAt(0)!, channel.admin);
+    return !isInvalidChain;
+  }
+
+  private checkVoteCompletedPositive(vote: VoteChain, admin: Identity): boolean {
+    if (
```

Comment:
since this method is _only_ a condition, maybe just the condition could be returned?
```ts
return (
  // everything currently in the if condition space.
);
```

Reply:
Fixed in eeb6ff1